src/libraries/Microsoft.PowerFx.Core/Public/Values/TableValue.cs

```diff
@@ -99,30 +99,72 @@ public virtual int Count()
         /// <summary>
         /// Lookup the record at the given 1-based index, or return an error value if out of range.
         /// </summary>
-        /// <param name="index1">1-based index.</param>
+        /// <param name="index1">1-based index.</param>
+        /// <param name="mutationCopy">copies the element, and the table entry pointing to it, when in a mutation context.</param>
         /// <returns>The record or an errorValue.</returns>
-        public DValue<RecordValue> Index(int index1)
+        public DValue<RecordValue> Index(int index1, bool mutationCopy = false)
         {
-            if (TryGetIndex(index1, out var record))
+            if (TryGetIndex(index1, out var record, mutationCopy: mutationCopy))
             {
                 return record;
             }
 
             return DValue<RecordValue>.Of(ArgumentOutOfRange(IRContext));
         }
 
-        // Index() does standard error messaging and then call TryGetIndex().
-        protected virtual bool TryGetIndex(int index1, out DValue<RecordValue> record)
-        {
+        // Index() does standard error messaging and then call TryGetIndex().
+        // By default, can't mutate through a table, must be overrideen for specific table types.
+        protected virtual bool TryGetIndex(int index1, out DValue<RecordValue> record, bool mutationCopy = false)
+        {
             var index0 = index1 - 1;
-            if (index0 < 0)
+            if (index0 < 0 || mutationCopy)
```

Comment:
> || mutationCopy [](http://example.com/codeflow?start=26&length=16)

Why this?  

Is there any user expression that can hit this path? (does binder prevent this from happening). If the only way we can get here is is developer didn't implement the virtuals properly, should we throw a hard exception instead?

Reply:
Good idea, I've changed it to an exception (in the mutationCopy argument overload version).